src/dataflow/src/sink/kafka.rs

```diff
@@ -818,154 +910,99 @@ where
-                                    row_index,
-                                    repeat_counter,
-                                    total_sent,
-                                }
-                            }
-                        }
-                        SendState::End(total_count) => {
-                            if s.consistency.is_some() {
-                                let result = s.send_consistency_record(
-                                    &ts.to_string(),
-                                    "END",
-                                    Some(total_count),
-                                );
-
-                                if let Err(retry) = result {
-                                    return retry;
-                                }
-                            }
-                            SendState::CommitTxn
-                        }
-                        SendState::CommitTxn => {
-                            let result = if s.transactional {
-                                s.commit_transaction(s.txn_timeout).await
-                            } else {
-                                Ok(())
-                            };
-
-                            match result {
-                                Ok(()) => {
-                                    // sanity check for the continuous updating
-                                    // of the write frontier below
-
-                                    s.assert_progress(ts);
-                                    progress_update.replace(ts.clone());
-
-                                    s.ready_rows.pop_front();
-                                    SendState::BeginTxn
-                                }
-                                Err(e) => s.transition_on_txn_error(s.send_state, *ts, e),
-                            }
-                        }
-                        SendState::AbortTxn => {
-                            let result = if s.transactional {
-                                s.abort_transaction(s.txn_timeout).await
-                            } else {
-                                Ok(())
-                            };
-
-                            match result {
-                                Ok(()) => SendState::BeginTxn,
-                                Err(e) => s.transition_on_txn_error(s.send_state, *ts, e),
-                            }
-                        }
-                        SendState::Shutdown => {
-                            s.shutdown_flag.store(true, Ordering::SeqCst);
-                            // Indicate that the sink is closed to everyone else who
-                            // might be tracking its write frontier.
-                            info!("shutting down kafka sink: {}", &s.name);
-                            break;
-                        }
+            // Can't use `?` when the return type is a `bool` so wrap up all the tx error handling
+            macro_rules! retry_txn_with_error {
```

Comment:
Hm, this looks a bit too specific. How about:

```suggestion
// Can't use `?` when the return type is a `bool` so define a custom try operator
macro_rules! bail_err {
    ($expr:expr) => {
        match $expr {
            Ok(val) => val,
            Err(_) => return true,
        }
    }
}
```